docs/puzzles/singletons.md

```diff
@@ -0,0 +1,437 @@
+Instead, what we can use is a **launcher** which is a specific puzzle that does exactly one thing: create a single singleton.  We then need to curry this launcher puzzle hash into our singleton and have the first singleton assert that it, in fact, came from a parent whose puzzle hash was the launcher puzzle hash.  Then, when people look at our singleton, they can see that the launcher puzzle hash is the hash of what they know to be a puzzle that creates only one singleton.  They don't need to go back to the original parent and verify because the singleton puzzle takes care of that right from the start!
+
+So what does the launcher look like?  Here's the source:
+
+```chialisp
+(mod (singleton_full_puzzle_hash amount key_value_list)
+
+  (include condition_codes.clvm)
+
+  (defun sha256tree (TREE)
+      (if (l TREE)
+          (sha256 2 (sha256tree (f TREE)) (sha256tree (r TREE)))
+          (sha256 1 TREE)
+      )
+  )
+
+  ; main
+  (list (list CREATE_COIN singleton_full_puzzle_hash amount)
+        (list CREATE_COIN_ANNOUNCEMENT (sha256tree (list singleton_full_puzzle_hash amount key_value_list))))
+)
+```
+
+Essentially two lines, so not too bad right?  One of the first things you may notice is that we don't curry anything in.  We actually cannot curry anything in because we want this puzzle hash to be constant among all singletons.  That way, even if someone isn't familiar with us, they know that if we came from this specific launcher puzzle hash, we can be trusted to be a unique singleton.
+
+For the most part, you simply put in `CREATE_COIN` parameters and the puzzle creates the singleton for you.  The tricky part is the announcement creation.  Since these parameters are not curried in, we somehow need them to be immune from the manipulations of malicious full nodes.  We cannot curry in a pubkey to sign them, or else our puzzle hash is no longer static. Our solution to this conundrum is to create an announcement from this puzzle that its parent asserts in the same block.  Usually, the parent is going to be a standard coin. In the standard coin, we sign the puzzle that makes the conditions.  If we create an `ASSERT_COIN_ANNOUNCEMENT` condition, we implicitly sign that too. That means we can implicitly sign all of the launcher solution values through asserting this announcement.  If any of those values are changed, the coin that creates the launcher will fail and thus the launcher will never be created!
+
+The last thing to note is the seemingly useless `key_value_list` that is passed in as an argument and announced.  The purpose for this is mostly external.  Sometimes you want to be able to know information about a puzzle before it is revealed.  The only way we can get this information on chain is from the parent's puzzle reveal so sometimes it is useful to have useless parameters be part of the solution in order to make it easier to follow the puzzle's on chain state.  Remember that you pay cost for every byte though so keep it concise.
```

Comment:
Maybe "... to communicate information to blockchain observers"

Reply:
"It is useful to have useless parameters" is a little contradictory... maybe "It is useful to have parameters that act as comments" or something.

Reply:
Might be worth describing the key-value pair mechanism that we're trying to standardize on and the rational for that (which is, standardize parsing).